src/game/common/system/upgrade.cpp

```diff
@@ -102,3 +117,140 @@ bool UpgradeCenter::Can_Afford_Upgrade(Player *player, const UpgradeTemplate *up
 
     return false;
 }
+
+UpgradeTemplate *UpgradeCenter::New_Upgrade(const Utf8String &name)
+{
+    UpgradeTemplate *upgrade = new UpgradeTemplate();
+    const UpgradeTemplate *default_upgrade = Find_Upgrade("DefaultUpgrade");
+
+    if (default_upgrade != nullptr) {
+        *upgrade = *default_upgrade;
+    }
+
+    upgrade->Set_Name(name);
+    upgrade->Set_Name_Key(g_theNameKeyGenerator->Name_To_Key(name.Str()));
+    BitFlags<128> mask;
+    mask.Set(m_upgradeCount, true);
+    m_upgradeCount++;
+    captainslog_dbgassert(m_upgradeCount < 128, "Can't have over %d types of Upgrades and have a Bitfield function.", 128);
+    upgrade->Set_Upgrade_Mask(mask);
+    Link_Upgrade(upgrade);
+    return upgrade;
+}
+
+void UpgradeCenter::Link_Upgrade(UpgradeTemplate *upgrade)
+{
+    if (upgrade != nullptr) {
+        upgrade->Friend_Set_Prev(nullptr);
+        upgrade->Friend_Set_Next(m_upgradeList);
+
+        if (m_upgradeList != nullptr) {
+            m_upgradeList->Friend_Set_Prev(upgrade);
+        }
+
+        m_upgradeList = upgrade;
+    }
+}
+
+void UpgradeCenter::Unlink_Upgrade(UpgradeTemplate *upgrade)
+{
+    if (upgrade != nullptr) {
+        if (upgrade->Friend_Get_Next() != nullptr) {
+            upgrade->Friend_Get_Next()->Friend_Set_Prev(upgrade->Friend_Get_Prev());
+        }
+
+        if (upgrade->Friend_Get_Prev() != nullptr) {
+            upgrade->Friend_Get_Prev()->Friend_Set_Next(upgrade->Friend_Get_Next());
+        } else {
+            m_upgradeList = upgrade->Friend_Get_Next();
+        }
+    }
+}
+
+std::vector<Utf8String> UpgradeCenter::Get_Upgrade_Names()
+{
+    std::vector<Utf8String> names;
+
+    for (UpgradeTemplate *tmplate = m_upgradeList; tmplate != nullptr; tmplate = tmplate->Friend_Get_Next()) {
+        names.push_back(tmplate->Get_Name());
+    }
+
+    return names;
+}
+
+void UpgradeCenter::Parse_Upgrade_Definition(INI *ini)
+{
+    Utf8String str = ini->Get_Next_Token();
+    UpgradeTemplate *upgrade =
+        g_theUpgradeCenter->Find_Non_Const_Upgrade_By_Key(g_theNameKeyGenerator->Name_To_Key(str.Str()));
+
+    if (upgrade == nullptr) {
+        upgrade = g_theUpgradeCenter->New_Upgrade(str);
+    }
+
+    captainslog_dbgassert(upgrade != nullptr, "Parse_Upgrade_Definition: Unable to allocate upgrade '%s'", str.Str());
+    ini->Init_From_INI(upgrade, UpgradeTemplate::Get_Field_Parse());
+}
+
+UpgradeTemplate::UpgradeTemplate() :
+    m_type(UPGRADE_TYPE_PLAYER),
+    m_nameKey(NAMEKEY_INVALID),
+    m_buildTime(0.0f),
+    m_cost(0),
+    m_academyClassify(ACT_NONE),
+    m_next(nullptr),
+    m_prev(nullptr),
+    m_buttonImage(nullptr)
+{
+}
+
+int UpgradeTemplate::Calc_Time_To_Build(Player *player) const
+{
+#ifdef GAME_DEBUG_STRUCTS
+    if (player->Is_Instant_Build()) {
+        return 1;
+    } else
+#endif
+    {
+        return m_buildTime * 30.0f;
```

Comment:
This returns `float` on `int`. As designed?

Reply:
This is correct to the original.